pkg/changelog/changelog.go

```diff
@@ -1,224 +1,204 @@
+type Entries[T comparable] struct {
+	entryFlagsMap EntryFlagsMap // Configuration map defining flags for each member kind.
+	entries       []entry[T]    // List of recorded entries.
 }
 
-// Get returns the value of the changelog at the given time.
-func (clv *Changelog[T]) Get(targetTime time.Time) T {
-	if len(clv.changes) == 0 {
-		return returnZero[T]()
+// NewEntries initializes a new `Entries` structure using the provided `EntryFlagsMap`.
+func NewEntries[T comparable](m EntryFlagsMap) *Entries[T] {
+	for kind, maxEntries := range m {
+		if maxEntries == 0 {
+			panic(fmt.Sprintf("kind %d must have MaxEntries > 0", kind))
+		}
 	}
 
-	idx := clv.findIndex(targetTime)
-	if idx == 0 {
-		return returnZero[T]()
+	return &Entries[T]{
+		entryFlagsMap: m,
+		entries:       []entry[T]{},
 	}
-
-	return clv.changes[idx-1].value
 }
 
-// GetAll returns all the values of the changelog.
-func (clv *Changelog[T]) GetAll() []T {
-	values := make([]T, 0, len(clv.changes))
-	for _, change := range clv.changes {
-		values = append(values, change.value)
+// Set adds or updates an entry in the Entries for the specified `MemberKind` ordered by timestamp.
+// If the new entry has the same value as the latest one, only the timestamp is updated.
+// If there are already the maximum number of entries for this kind, it reuses or replaces an existing entry.
+//
+// ATTENTION: Make sure to pass a value of the correct type for the specified `MemberKind`.
+func (e *Entries[T]) Set(k MemberKind, value T, t time.Time) {
+	maxEntries, ok := e.entryFlagsMap[k]
+	if !ok {
+		panic(fmt.Sprintf("kind %d is not present in the entryFlagsMap", k))
 	}
-	return values
-}
 
-// Setters
+	maxSize := int(maxEntries)
+	indexes := make([]int, 0)
 
-// SetCurrent sets the latest value of the changelog.
-func (clv *Changelog[T]) SetCurrent(value T) {
-	clv.setAt(value, time.Now())
-}
-
-// Set sets the value of the changelog at the given time.
-func (clv *Changelog[T]) Set(value T, targetTime time.Time) {
-	clv.setAt(value, targetTime)
-}
-
-// private
-
-// setAt sets the value of the changelog at the given time.
-func (clv *Changelog[T]) setAt(value T, targetTime time.Time) {
-	// If the timestamp is already set, update that value only.
-	_, ok := clv.timestamps[targetTime]
-	if ok {
-		index := clv.findIndex(targetTime) - 1
-		if index < 0 {
-			logger.Debugw("changelog internal error: illegal index for existing timestamp")
+	// collect indexes of entries equal to kind
+	for idx, entry := range e.entries {
+		if entry.k == k {
+			indexes = append(indexes, idx)
 		}
-		if !clv.changes[index].timestamp.Equal(targetTime) { // sanity check only (time exists already)
-			logger.Debugw("changelog internal error: timestamp mismatch")
+	}
+
+	// if there are entries for kind check if the last entry has the same value
+	if len(indexes) > 0 {
+		lastIdx := indexes[len(indexes)-1]
+		if e.entries[lastIdx].value == value && t.After(e.entries[lastIdx].t) {
```

Comment:
This is basically a case of keeping the timestamp warm. Is this a documented need?

Reply:
Actually this is a shortcircuit to avoid the previous coalesce logic. Instead of always inserting the value, even if it's equal to later enforce the boundaries and remove the old ones, the shortcircuit always checks if it's an update or an insertion always keeping the slice tidy.